crates/bevy_render/src/shader/shader_defs.rs

```diff
@@ -59,56 +53,180 @@ impl ShaderDef for Option<Handle<Texture>> {
     }
 }
 
+#[derive(Debug, Clone, Hash, Eq, PartialEq, Reflect)]
+pub enum ShaderDefSource {
+    Component(Uuid),
+    Asset(HandleUntyped),
+}
+
+impl<T: Asset> From<&Handle<T>> for ShaderDefSource {
+    fn from(h: &Handle<T>) -> Self {
+        Self::Asset(h.clone_weak_untyped())
+    }
+}
+
+impl From<Uuid> for ShaderDefSource {
+    fn from(uuid: Uuid) -> Self {
+        Self::Component(uuid)
+    }
+}
+
 /// Updates [RenderPipelines] with the latest [ShaderDefs]
-pub fn shader_defs_system<T>(mut query: Query<(&T, &mut RenderPipelines)>)
+pub fn shader_defs_system<T>(mut query: Query<(&T, &mut RenderPipelines), Changed<T>>)
+where
+    T: ShaderDefs + TypeUuid + Send + Sync + 'static,
+{
+    query
+        .iter_mut()
+        .map(|(s, p)| (s, (T::TYPE_UUID).into(), p))
```

Comment:
Can we use the TypeId here instead of TypeUuid? So far we aren't requiring TypeUuid on components and we use TypeId as the unique identifier everywhere else.

Reply:
(Assets do require TypeUuid though)